gms/feature_service.hh

```diff
@@ -64,6 +64,7 @@ public:
     explicit feature_service(feature_config cfg);
     ~feature_service() = default;
     future<> stop();
+    future<> enable(std::set<std::string_view> list, db::system_keyspace& sys_ks);
```

Comment:
I don't see the need to tune feature_service API to allow for persistent and non-persistent enabling. Current code allows you both already. If you need to enable features without persisting, use feature_service::enable(), if you need to persist enabling -- get the persistent enabler and use _it_.

Reply:
@xemul I needed to expose the possibility of enabling features persistently from outside (from raft group 0). I don't see how to cleanly expose the persistent enabler, could you tell what you had in mind?

The persistent enabler seems to be designed to be a listener, perhaps I could add a way to register callbacks that listen to changes to the group 0 state. I'm not sure if I like that idea, I think features should be enabled explicitly there.

Besides, I think that the current `feature_service::enable` API is a bit unsafe. Enabling features without persisting them first is not safe in general, and the fact that this method is called `feature_service::enable` and is public makes it too tempting to use it. I think the API should be more explicit about what it does, which is what I tried to do with this cleanup.

(Another besides: I'll move the changes to the feature service to a separate PR, but let's first clear the issues here first)

Reply:
@xemul ping

Reply:
Sorry

> @xemul I needed to expose the possibility of enabling features persistently from outside (from raft group 0). I don't see how to cleanly expose the persistent enabler, could you tell what you had in mind?

In that case you need to patch persistent enabler, not feature service. The latter is nowadays a "transport" that spreads features enabling around subscribers

> The persistent enabler seems to be designed to be a listener, perhaps I could add a way to register callbacks that listen to changes to the group 0 state. I'm not sure if I like that idea, I think features should be enabled explicitly there.

Not that it's _designed_ to be a listener. There was just one source of features-to-be-enabled and it was gossiper. Now it's raft, so feel free to patch this enabler respectively

> Besides, I think that the current feature_service::enable API is a bit unsafe. Enabling features without persisting them first is not safe in general, and the fact that this method is called feature_service::enable and is public makes it too tempting to use it. I think the API should be more explicit about what it does, which is what I tried to do with this cleanup.

It's not supposed to be safe, because, as I mentioned above, feature_service is just a way to tell the rest of the system about enabling features. It's been such in the old days, then it was considered that not persisting features was unsafe (a2f5ad432f657253a3a9475be152ee52c05ce30f) and it was done by patching feature-service introducing nasty dependency loops via ugly `qctx` thing. I propose not to walk this route for the 2nd time

> (Another besides: I'll move the changes to the feature service to a separate PR, but let's first clear the issues here first)

:+1: